src/net/ossl.cc

```diff
@@ -859,13 +859,119 @@ class session : public enable_shared_from_this<session>, public session_impl {
         return make_ready_future<result_t>(std::move(dn));
     }
 
-    future<std::vector<subject_alt_name>> get_alt_name_information(std::unordered_set<subject_alt_name_type>) override {
+    future<std::vector<subject_alt_name>> get_alt_name_information(std::unordered_set<subject_alt_name_type> types) override {
         using result_t = std::vector<subject_alt_name>;
-        return make_exception_future<result_t>(std::runtime_error("unimplemented"));
+
+        if (_error) {
+            return make_exception_future<result_t>(_error);
+        }
+        if (_shutdown) {
+            return make_exception_future<result_t>(std::system_error(ENOTCONN, std::system_category()));
+        }
+        if (!connected()) {
+            return handshake().then([this, types = std::move(types)]() mutable {
+               return get_alt_name_information(std::move(types));
+            });
+        }
+
+        auto peer_cert = get_peer_certificate();
+        if (!peer_cert) {
+            return make_ready_future<result_t>();
+        }
+        return make_ready_future<result_t>(do_get_alt_name_information(peer_cert, types));
     }
 
 
 private:
+    std::vector<subject_alt_name> do_get_alt_name_information(std::shared_ptr<X509> &peer_cert,
```

Comment:
nit: I'd be tempted to invert all the conditions and early-exit to reduce the nesting in this function. You can probably name the return early and get NRVO.